(DOCS-15613): Kotlin Migration documentation (#2514)
mongodb/docs-realm: source/sdk/kotlin/realm-database/change-an-object-model.txt

   flag to delete the database automatically when a schema mismatch would
   require a migration. 

   Never release an app to production with this flag set to ``true``.

.. _kotlin-schema-version:

Schema Version
--------------

A **schema version** identifies the state of a 
realm schema at some point in time. 
Realm Database tracks the schema version of each realm and uses it to map the
objects in each realm to the correct schema.

Schema versions are ascending integers that you can optionally include
in the realm configuration when you :ref:`open a realm <kotlin-open-a-realm>`. 
If a client application does not specify a version number when it opens a 
realm, then the realm defaults to version ``0``.

.. important:: Increment Versions Monotonically
   
   Migrations must update a realm to a higher schema version. 
   Realm Database will throw an error if a client
   application opens a realm with a schema version that is lower than
   the realm's current version or if the specified schema version is the
   same as the realm's current version but includes different
   object schemas.

.. _kotlin-migrations:

Migration

> ---------

Reviewer: ```suggestion
---------

```

Schema updates are called **migrations** in Realm Database.
A migration updates a realm and any objects it
contains from one schema version to a newer version. 

Whenever you open an existing realm with a schema version greater than the
realm's current version, you can provide a **migration function** that 
defines any additional logic needed for the schema update.
For example: 

- Setting property values
- Combining or splitting fields
- Renaming a field
- Changing a field's type
  
The function has access to the realm's version number and incrementally
updates objects in the realm to conform to the new schema.

A **local migration** is a migration for a realm that does *not* 
automatically :ref:`Sync <sync>` with another realm. 

> .. note::
>       To modify schema properties of a synced realm, refer to 
>       :ref:`Update a Data Model <synced-schema-overview>` in the Atlas
>       Device Sync documentation.

Reviewer: ```suggestion
.. note::

   To modify schema properties of a synced realm, refer to 
   :ref:`Update a Data Model <synced-schema-overview>` in the Atlas
   Device Sync documentation.
```

Automatically Update Schema
---------------------------

If your schema update adds or removes properties, 
Realm Database can perform the migration automatically. 
You only need to increment the ``schemaVersion``. 

.. _kotlin-add-a-property:

Add a Property
~~~~~~~~~~~~~~

To add a property to a schema:  

> #. Add the new property to the object's class. 

Reviewer: ```suggestion
#. Add the new property to the ``RealmObject`` definition. 
```

nit: for precision. feel free to adjust phrasing

#. Set the schema version through the 
   `RealmConfiguration <{+kotlin-local-prefix+}io.realm.kotlin/-realm-configuration/index.html>`__ builder.

Realm does not automatically set values for new properties 

> unless the updated object schema specifies a default value. 

Reviewer: does Realm automatically set values for new properties of existing if the new property in updated object schema contains a default value? 

i think this sentence could be rephrased to make clearer what the behavior is here.

Author: clarified

migrate old instances of a given object to the new schema.

When you open the realm with the updated schema, you must do the following
in the 
`RealmConfiguration <{+kotlin-local-prefix+}io.realm.kotlin/-realm-configuration/index.html>`__:

- Increment the ``schemaVersion`` property.
- Define the migration logic using the `migrationContext <{+kotlin-local-prefix+}io.realm.kotlin.migration/-automatic-schema-migration/-migration-context/index.html>`__.

.. _kotlin-modify-a-property:
.. _kotlin-rename-a-property:

Modify a Property
~~~~~~~~~~~~~~~~~

To modify an object property (e.g. rename, merge, split, or change property
type): 

#. Change the property or properties in the object schema.
#. Open the realm with an incremented schema version and a
   migration function that maps the existing objects to use the new properties.

In the following example, the schema is updated to change a property type, merge two properties into a new property, and rename an existing property. 

.. literalinclude:: /examples/generated/kotlin/MigrationTest.snippet.model-v4.kt
   :language: kotlin

.. literalinclude:: /examples/generated/kotlin/MigrationTest.snippet.local-migration.kt

>    :language: kotlin

Reviewer: generally, we don't stack code examples like this w/o text in between. could you split the sentence into 2, one introducing each of the code examples?

Author: fixed

.. _kotlin-convert-to-an-embedded-object:

> .. note:: 

Reviewer: ```suggestion
.. note:: 

```

  If your schema update includes converting an object to an 

>   `EmbeddedObject <{+kotlin-local-prefix+}io.realm.kotlin.types/-embedded-realm-object/index.html>`__, 

Reviewer: what do you mean by 'convert an object to an embedded object'?

Author: Exactly that: as part of your schema update, you change a `RealmObject` to an `EmbeddedRealmObject`. It was an explicit section in the Swift doc which is why I included it here as a callout. Updated for clarity.

  the migration function must ensure that every embedded object has exactly 

>   one backlink to a parent object. Embedded objects cannot exist independently of a parent object. 

Reviewer: i'm a bit confused why backlinks are mentioned here. backlinks aren't generally needed to use embedded objects (at least in the other SDKs)

Author: Tweaked wording

.. _kotlin-other-migration-tasks:

Other Migration Tasks
~~~~~~~~~~~~~~~~~~~~~

To perform other realm schema migrations, use the following properties of
the ``AutomaticSchemaMigration.MigrationContext`` interface:

- `oldRealm <{+kotlin-local-prefix+}io.realm.kotlin.migration/-automatic-schema-migration/-migration-context/old-realm.html>`__: 

>   The Realm as it existed before the migration with the previous schema 

Reviewer: ```suggestion
  The realm as it existed before the migration with the previous schema 
```

i think we want lowercase b/c that's how we refer to an individual realm database instance.

Reviewer: yeah it should be 'realm'. as a general rule, use lowercase 'realm' when referring to an instance of realm database.

if you wouldn't mind updating the flutter docs, i would appreciate that 😄

  version.
  The `dynamic API <{+kotlin-local-prefix+}io.realm.kotlin.dynamic/index.html>`__ lets you find Realm objects by a string representation of 
  their class name.
- `newRealm <{+kotlin-local-prefix+}io.realm.kotlin.migration/-automatic-schema-migration/-migration-context/new-realm.html>`__:

>   The Realm as it exists after the migration using the new schema version. 

Reviewer: ```suggestion
  The realm as it exists after the migration using the new schema version. 
```

ditto https://github.com/mongodb/docs-realm/pull/2514/files#r1084640032

> By the end of the migration, you must migrate all relevant data from ``oldRealm`` into ``newRealm``. 

Reviewer: what do you mean by 'relevant data' here? 

do you mean 'data that you want to be migrated'? what happens if you don't include in the migrated version.

Author: that wording was also shamelessly stolen from the Flutter doc 😅 I tweaked to clarify a bit but I'm waiting on confirmation from sdk team on the answer to your question.

Author: Updated for clarity

Reviewer: i like the edits! if you wouldn't mind applying that wording to flutter while you're decapitalizing 'realm', i'd appreciate it!

> Any objects obtained in the migration are valid only in the scope of the migration.

Reviewer: what's does 'objects obtained in the migration' mean?
